File: Secure Code Review/SSRF/Java/ssrf.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@WebServlet("/fetchData")
public class SSRFServlet extends HttpServlet {

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        String targetUrl = request.getParameter("url"); // User-controlled input
        
        if (targetUrl == null || targetUrl.isEmpty()) {
            response.getWriter().println("No URL provided!");
            return;
        }

        try {
                                                                                                                                            // Directly use the user-provided URL to make a request
            URL url = new URL(targetUrl);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");

            BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            StringBuilder responseContent = new StringBuilder();
            String inputLine;

            while ((inputLine = in.readLine()) != null) {
                responseContent.append(inputLine);
            }
            in.close();

            // Return the response from the requested URL
            response.getWriter().println(responseContent.toString());
        } catch (Exception e) {
            response.getWriter().println("Error fetching URL: " + e.getMessage());
        }
    }
}



















............................................................................................................................................
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


How This above snipped Code is Vulnerable to SSRF

    User-Controlled Input for URL:
        The targetUrl parameter is directly used to make an HTTP request without validation or restrictions.
        An attacker can provide any URL, including internal/private network endpoints.

    Exploitation:
        Access Internal Resources: An attacker could access internal systems (e.g., http://localhost:8080/admin) that are not directly exposed to the internet.
        Leverage Private IPs: Access internal services like http://192.168.1.1 or http://127.0.0.1 to exploit vulnerable applications or services.
        Trigger Data Exfiltration: Combine SSRF with other vulnerabilities to exfiltrate sensitive data.

    Example Exploit:
        URL: http://yourserver.com/fetchData?url=http://localhost:8080/admin
        The attacker can access sensitive internal endpoints or services.

File: Secure Code Review/SSRF/.Net/ssrf.cs
using System;
using System.IO;
using System.Net;
using Microsoft.AspNetCore.Mvc;

namespace VulnerableApp.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class FetchController : ControllerBase
    {
        [HttpGet]
        public IActionResult GetExternalData([FromQuery] string url)
        {
            if (string.IsNullOrEmpty(url))
            {
                return BadRequest("URL parameter is required.");
            }

            try
            {
                // Directly use the user-provided URL to make an HTTP request
                var request = WebRequest.Create(url);
                using (var response = request.GetResponse())
                using (var reader = new StreamReader(response.GetResponseStream()))
                {
                    var data = reader.ReadToEnd();
                    return Ok(data);
                }
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error fetching data: {ex.Message}");
            }
        }
    }
}

















////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
............................................................................................................................................


  How This snipped Code is Vulnerable

    User-Controlled Input:
        The url parameter is taken directly from the query string and passed to WebRequest.Create() without validation or restrictions.
        Example: An attacker could call the endpoint like this:

    GET /api/fetch?url=http://localhost:8080/admin

Impact:

    Access Internal Resources: The attacker could access internal/private network resources that are not meant to be exposed to the internet.
    Exfiltrate Data: If the internal service returns sensitive data, it can be leaked to the attacker.
    Leverage Internal Services: Attackers can interact with internal APIs or services, potentially triggering unintended actions.

Example Exploit:

    A crafted request to http://localhost:5000/api/fetch?url=http://127.0.0.1:6379 could allow the attacker to probe internal services (e.g., Redis, AWS metadata service).








How to Fix This Vulnerability

    Validate the Input URL:
        Ensure the URL adheres to a predefined format or allowlist of domains.

    Restrict Access to Internal IPs:
        Block requests to private IP ranges or loopback addresses.

    Use a Safe HTTP Client:
        Use HttpClient with proper validation and constraints instead of directly working with WebRequest.















